src/rest-server/src/models/v2/job.js

```diff
@@ -0,0 +1,297 @@
+      name: taskRole,
+      taskNumber: frameworkDescription.taskRoles[taskRole].taskNumber,
+      cpuNumber: frameworkDescription.taskRoles[taskRole].taskService.resource.cpuNumber,
+      memoryMB: frameworkDescription.taskRoles[taskRole].taskService.resource.memoryMB,
+      gpuNumber: frameworkDescription.taskRoles[taskRole].taskService.resource.gpuNumber,
+      shmMB: 512,
+      minFailedTaskCount: frameworkDescription.taskRoles[taskRole].applicationCompletionPolicy.minFailedInstances,
+      minSucceededTaskCount: frameworkDescription.taskRoles[taskRole].applicationCompletionPolicy.minSucceededInstances,
+    },
+    tasksNumber: tasksNumber,
+    taskRoleList: Object.keys(frameworkDescription.taskRoles).join(','),
+    taskRolesNumber: Object.keys(frameworkDescription.taskRoles).length,
+    jobEnvs: null,
+    hdfsUri: launcherConfig.hdfsUri,
+    aggregatedStatusUri: launcherConfig.frameworkAggregatedStatusPath(frameworkName),
+    frameworkInfoWebhdfsUri: launcherConfig.frameworkInfoWebhdfsPath(frameworkName),
+    azRDMA: azureEnv.azRDMA === 'true' ? true : false,
+    reqAzRDMA: false,
+    inspectPidFormat: '{{.State.Pid}}',
+    inspectOOMKilledFormat: '{{.State.OOMKilled}}',
+  });
+  return yarnContainerScript;
+};
+
+// backward compatible generation for docker container script template in api v1
+const generateDockerContainerScript = (frameworkName, userName, config, taskRole) => {
+  const dockerContainerScript = mustache.render(dockerContainerScriptTemplate, {
+    idx: taskRole,
+    jobData: {
+      jobName: frameworkName,
+      userName: userName,
+    },
+    taskData: {
+      command: config.taskRoles[taskRole].entrypoint,
+    },
+    hdfsUri: launcherConfig.hdfsUri,
+    webHdfsUri: launcherConfig.webhdfsUri,
+    paiMachineList: paiConfig.machineList,
+    azRDMA: azureEnv.azRDMA === 'true' ? true : false,
+    reqAzRDMA: false,
+  });
+  return dockerContainerScript;
+};
+
+const generateSSHKeys = (frameworkName) => {
+  return new Promise((resolve) => {
+    keygen({
+      location: frameworkName,
+      comment: `ssh key for ${frameworkName}`,
+      read: true,
+      destroy: true,
+    }, (err, out) => {
+      if (err) {
+        logger.warn('Generating ssh key files failed! Will skip generating ssh info.');
+        resolve(null);
+      } else {
+        const keys = {};
+        // private key
+        keys[frameworkName] = out.key;
+        // public key
+        keys[`${frameworkName}.pub`] = out.pubKey;
+        resolve(keys);
+      }
+    });
+  });
+};
+
+const prepareContainerScripts = async (frameworkName, userName, config, rawConfig) => {
+  // hdfs operations
+  const hdfs = new HDFS(launcherConfig.webhdfsUri);
+  // async mkdir on hdfs
+  const mkdir = async (path, user, mode) => {
+    const options = {'user.name': user, 'permission': mode};
+    return new Promise((resolve) => {
+      hdfs.createFolder(path, options, (err, result) => {
+        if (err) {
+          throw err;
```

Comment:
You cannot use `throw` here since it is a callback but not directly in Promise contructor.

[`require('util').promisify`](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original) is your friend.

Reply:
updated